accepted/2021/statics-in-interfaces/README.md

```diff
@@ -8,6 +8,793 @@ C# is looking at enabling static abstract members in interfaces (https://github.
+    }
+
+    public interface ISpanParseable<TSelf> : IParseable<TSelf>
+        where TSelf : ISpanParseable<TSelf>
+    {
+        // This inherits from IParseable, much as ISpanFormattable is planned to inherit from IFormattable
+
+        static abstract TSelf Parse(ReadOnlySpan<char> s, IFormatProvider? provider);
+
+        static abstract bool TryParse(ReadOnlySpan<char> s, IFormatProvider? provider, out TSelf result);
+    }
+
+    public interface IEquatableOperators<TSelf, TOther> : IEquatable<TOther>
+        where TSelf : IEquatableOperators<TSelf, TOther>
+    {
+        // Given this takes two generic parameters, we could simply call it IEquatable<TSelf, TOther>
+
+        // This should not be `IEquatableOperators<TSelf>` as there are types like `Complex` where
+        // TOther can be `double` and represent an optimization, both in size and perf
+
+        static abstract bool operator ==(TSelf lhs, TOther rhs);
+
+        static abstract bool operator !=(TSelf lhs, TOther rhs);
+    }
+
+    public interface IComparableOperators<TSelf, TOther> : IComparable<TOther>, IEquatableOperators<TSelf, TOther>
+        where TSelf : IComparableOperators<TSelf, TOther>
+    {
+        // Given this takes two generic parameters, we could simply call it IComparable<TSelf, TOther>
+
+        // This inherits from IEquatableOperators<TSelf, TOther> as even though IComparable<T> does not
+        // inherit from IEquatable<T>, the <= and >= operators as well as CompareTo itself imply equality
+
+        static abstract bool operator <(TSelf lhs, TOther rhs);
+
+        static abstract bool operator <=(TSelf lhs, TOther rhs);
+
+        static abstract bool operator >(TSelf lhs, TOther rhs);
+
+        static abstract bool operator >=(TSelf lhs, TOther rhs);
+    }
+
+    public interface IAddable<TSelf, TOther>
+        where TSelf : IAddable<TSelf, TOther>
+    {
+        // An additive identity is not exposed here as it may be problematic for cases like
+        // DateTime that implements both IAddable<DateTime, DateTime> and IAddable<DateTime, TimeSpan>
+        // This would require overloading by return type and may be problematic for some scenarios where TOther might not have an additive identity
+
+        // We can't assume TOther + TSelf is valid as not everything is commutative
+
+        // We can't expose TSelf - TOther as some types only support one of the operations
+
+        static abstract TSelf operator +(TSelf value);
+
+        static abstract TSelf operator +(TSelf lhs, TOther rhs);
+
+        // There is an open question on if we should support "checked" versions of the operators
+        // Many algorithms, such as in LINQ, depend on checked behavior for overflow
+        // Supporting this would require well-defined ECMA-335 names and language work to ensure the right things happen
+    }
+
+    public interface ISubtractable<TSelf, TOther>
```

Comment:
A little bit surprised by the types here.  I would have expected

```csharp
public interface ISubtractable<TSelf, TDiff>
  where TSelf : ISubtractable<TSelf, TDiff>
{
    static abstract TDiff operator- (TSelf lhs, TSelf rhs);
}
```

that is, a generalization of `DateTime dt1, dt2; TimeSpan diff = dt1 - dt2`, rather than of `DateTime dt1; TimeSpan diff;  DateTime dt2 = dt1 - diff` since the latter can be expressed in terms of addition and negation.

Reply:
The current design has been that `typeof(TSelf) == typeof(TResult)` because that greatly simplifies things in terms of consumption and matches what the vast majority of types end up following.

There are some interesting edge cases where `TResult` can differ, but that's a minority compared to where `TOther` can differ, and so we have an option of `TSelf, TOther` for simplicity or `TSelf, TOther, TResult` for completeness.
It may be that we do need `TSelf, TOther, TResult` in practice, especially as community feedback comes in and outreach to major libraries happens.

Reply:
I assume this means that people couldn't call do `dateTime1 - dateTime2` from generic code at all, right? Or is there a viable, but slightly clunky way to do that?

Reply:
Correct, we would need to expose `ISubtractable<TSelf, TOther, TResult>` where `DateTime : ISubtractable<DateTime, DateTime, TimeSpan>`

There may be enough scenarios that its worth doing this, even if its relatively uncommon.

Reply:
I've changed this to have `ISubtractable<TSelf, TOther, TResult>` as there are enough use cases in System that I think it justifies doing it from day 1, and it ensures we always meet the needs, since its the exact subtraction signature anyways.